lib/host_remove_duplicates.py

```diff
@@ -0,0 +1,166 @@
+from sqlalchemy import and_
+from sqlalchemy import not_
+from sqlalchemy import or_
+
+from app.models import Host
+from lib.metrics import delete_duplicate_host_count
+
+# from app.logging import get_logger
+
+# logger = None
+
+__all__ = ("delete_duplicate_hosts",)
+
+
+# The order is important, particularly the first 3 which are elevated facts with provider_id being the highest priority
+CANONICAL_FACTS = ("fqdn", "satellite_id", "bios_uuid", "ip_addresses", "mac_addresses")
+
+ELEVATED_CANONICAL_FACT_FIELDS = ("provider_id", "insights_id", "subscription_manager_id")
+
+
+def matches_at_least_one_canonical_fact_filter(canonical_facts):
+    # Contains at least one correct CF value
+    # Correct value = contains key:value
+    # -> OR( *correct values )
+    filter_ = ()
+    for key, value in canonical_facts.items():
+        filter_ += (Host.canonical_facts.contains({key: value}),)
+
+    return or_(*filter_)
+
+
+def contains_no_incorrect_facts_filter(canonical_facts):
+    # Does not contain any incorrect CF values
+    # Incorrect value = AND( key exists, NOT( contains key:value ) )
+    # -> NOT( OR( *Incorrect values ) )
+    filter_ = ()
+    for key, value in canonical_facts.items():
+        filter_ += (
+            and_(Host.canonical_facts.has_key(key), not_(Host.canonical_facts.contains({key: value}))),  # noqa: W601
+        )
+
+    return not_(or_(*filter_))
+
+
+def multiple_canonical_facts_host_query(canonical_facts, query):
+    query = query.filter(
+        (contains_no_incorrect_facts_filter(canonical_facts))
+        & (matches_at_least_one_canonical_fact_filter(canonical_facts))
+    )
+    return query
+
+
+# Get hosts by the highest elevated canonical fact present
+def find_host_by_elevated_canonical_facts(elevated_cfs, query, logger):
+    """
+    First check if multiple hosts are returned.  If they are then retain the with the highest
```

Comment:
Obviously not gonna block on this. Real nit-picking hours.

```suggestion
    First check if multiple hosts are returned.  If they are then retain the one with the highest
```